clang/lib/Analysis/ThreadSafety.cpp

```diff
@@ -876,14 +907,25 @@ class LockableFactEntry : public FactEntry {
 
   void handleLock(FactSet &FSet, FactManager &FactMan, const FactEntry &entry,
                   ThreadSafetyHandler &Handler) const override {
-    Handler.handleDoubleLock(entry.getKind(), entry.toString(), loc(),
-                             entry.loc());
+    if (std::unique_ptr<FactEntry> RFact = tryReenter(entry.kind())) {
+      // This capability has been reentrantly acquired.
+      FSet.replaceLock(FactMan, entry, std::move(RFact));
+    } else {
+      Handler.handleDoubleLock(entry.getKind(), entry.toString(), loc(),
+                               entry.loc());
+    }
   }
 
   void handleUnlock(FactSet &FSet, FactManager &FactMan,
                     const CapabilityExpr &Cp, SourceLocation UnlockLoc,
                     bool FullyRemove,
                     ThreadSafetyHandler &Handler) const override {
+    if (std::unique_ptr<FactEntry> RFact = leaveReentrant()) {
+      // This capability remains reentrantly acquired.
+      FSet.replaceLock(FactMan, Cp, std::move(RFact));
+      return;
+    }
+
     FSet.removeLock(FactMan, Cp);
     if (!Cp.negative()) {
       FSet.addLock(FactMan, std::make_unique<LockableFactEntry>(
```

Comment:
Interesting that both branches remove one lock and add another. I wonder if we shouldn't just do `FSet.removeLock(FactMan, Cp);` unconditionally and then add the other lock. (Side note: I don't know why we have `if (!Cp.negative())`, it seems weird to unlock a negative capability.)

The same would be true for `handleLock`, if we didn't remove the negative capability in `ThreadSafetyAnalyzer::addLock` instead.

Reply:
Simplifying this.